### PR TITLE
Fix several ztest failures

### PR DESCRIPTION
**Pull request by tcaputi**
This PR contains several patches which fix several common failures that buildbot is currently hitting when running `zloop.sh`. This patch also fixes a few long-standing bugs that prevented various debug output from being printed during ztest execution.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: cmd/ztest/ztest.c [hunk 1]
```diff
@@ -6652,13 +6652,20 @@ ztest_deadman_thread(void *arg)
...
+	delay = zs->zs_thread_stop - zs->zs_thread_start +
+	    MSEC2NSEC(zfs_deadman_synctime_ms);
 
-		(void) poll(NULL, 0, (int)NSEC2MSEC(delta));
+	while (!ztest_exiting) {
+		/*
+		 * Wait for the delay timer while checking occasionally
+		 * if we should stop.
+		 */
+		if (gethrtime() - last_run < delay) {
```
Comment (behlendorf) [inline cmd/ztest/ztest.c]: nit: personally I find this easier to grok, `if (gethrtime() > last_run + delay) { }`

File: cmd/ztest/ztest.c [hunk 2]
```diff
@@ -6687,7 +6694,11 @@ ztest_deadman_thread(void *arg)
 
 		(void) printf("ztest has been running for %lld seconds\n",
 		    total);
+
+		last_run = gethrtime();
```
Comment (behlendorf) [inline cmd/ztest/ztest.c]: Once we exceed the initial delay, I think it would be reasonable to set `delay = MSEC2NSEC(zfs_deadman_checktime_ms)`.

File: lib/libzpool/kernel.c [hunk 3]
```diff
@@ -783,7 +783,8 @@ dprintf_setup(int *argc, char **argv)
  * =========================================================================
  */
 void
-__dprintf(const char *file, const char *func, int line, const char *fmt, ...)
+__dprintf(boolean_t dprint, const char *file, const char *func,
+    int line, const char *fmt, ...)
```
[the remark this answers is not in the document]
Comment (tcaputi) [reply lib/libzpool/kernel.c]: @ahrens and I talked about this a bit. We wanted to differentiate between `dprintf()` and `zfs_dbgmsg()`. After talking for a bit we eventually decided that `dprintf()` should be printed immediately and `zfs_dbgmsg()` should be logged.

[the remark this answers is not in the document]
Comment (behlendorf) [reply lib/libzpool/kernel.c]: OK

File: cmd/zdb/zdb.c [hunk 4]
```diff
@@ -221,6 +221,7 @@ dump_debug_buffer(void)
 {
 	if (dump_opt['G']) {
 		(void) printf("\n");
+		(void) fflush(stdout);
 		zfs_dbgmsg_print("zdb");
```
Comment (behlendorf) [inline cmd/zdb/zdb.c]: nit: we should add an empty line after dumping the debug buffer for `zdb` and `ztest` to make it easier to read.  Right now, it's easy for the ASSERT to blend in to the previous log.  For example:

```
...
spa_misc.c:409:spa_load_note(): spa_load(ztest_CHECKPOINTED_UNIVERSE, config trusted): LOADED
spa.c:7098:spa_async_request(): spa=ztest_CHECKPOINTED_UNIVERSE async request task=32
spa_open(ztest_opts.zo_pool, &spa, FTAG) == 0 (0x6 == 0)
ASSERT at ztest.c:6834:ztest_run()/home/behlendo/src/git/zfs/cmd/ztest/.libs/lt-ztest[0x409eac]
/lib64/libpthread.so.0(+0xf6d0)[0x7f50873426d0]
/lib64/libc.so.6(gsignal+0x37)[0x7f5086f9c277]
/lib64/libc.so.6(abort+0x148)[0x7f5086f9d968]
/home/behlendo/src/git/zfs/cmd/ztest/.libs/lt-ztest[0x40aff1]
/home/behlendo/src/git/zfs/cmd/ztest/.libs/lt-ztest[0x408de7]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f5086f88445]
/home/behlendo/src/git/zfs/cmd/ztest/.libs/lt-ztest[0x409502]

ZFS_DBGMSG(ztest):
spa.c:4429:spa_open_common(): spa_open_common: opening ztest
spa_misc.c:409:spa_load_note(): spa_load(ztest, config trusted): LOADING
...
```

**Comment by sdimitro** [comment]
I'll give a review for this by tomorrow EOD the latest.

One quick question though:
From Brian's comment above it seems some of the code on this PR has made it to mainline. Can we update the PR so it is clear what problem we are trying to solve with this PR and also have the diffs showing only the new changes?

**Comment by behlendorf** [comment]
@sdimitro none of the changes in this PR have been integrated yet, but several of them do fix issues which were introduced by recent features.  In practice, I believe we've only encountered them with `ztest`.

The goal is to get `ztest` to the point where it can run indefinitely without encountering any unexpected failures.  The first 8 commits I referenced above improve the debug-ability and address several of the most commonly observed failures, but by no means all of them.

> also have the diffs showing only the new changes?

I'm not sure what you mean by this.  You can individually inspect the fixes by selecting each commit hash.  They've been grouped in to this PR to make testing easier, I agree we don't want to squash them when merging.  If it's helpful for review purposes I'm sure we can round up some examples of each of these failures.

File: cmd/ztest/ztest.c [hunk 5]
```diff
@@ -7384,7 +7384,7 @@ main(int argc, char **argv)
 	 * segments can be reconstructed in a reasonable amount of time
 	 * when reconstruction is known to be possible.
 	 */
-	zfs_reconstruct_indirect_damage_fraction = 4;
+	zfs_reconstruct_indirect_damage_fraction = 100;
```
Comment (ahrens) [inline cmd/ztest/ztest.c]: can you explain this a bit more?  why can't we do vdev_indirect_splits_damage() every time?  Why is 1/100th of the time OK?

File: module/zfs/vdev_indirect.c [hunk 6]
```diff
@@ -1632,7 +1632,7 @@ vdev_indirect_splits_damage(indirect_vsd_t *iv, zio_t *zio)
 		}
 
 		iv->iv_attempts_max *= 2;
-		if (iv->iv_attempts_max > (1ULL << 16)) {
+		if (iv->iv_attempts_max >= (1ULL << 12)) {
```
Comment (ahrens) [inline module/zfs/vdev_indirect.c]: is this related to the ztest fix? Or is this just making it run faster by not damaging when it would cause more than 2^12 tries?

File: cmd/ztest/ztest.c [hunk 7]
```diff
@@ -3529,13 +3528,9 @@ ztest_device_removal(ztest_ds_t *zd, uint64_t id)
 	uint64_t guid;
 	int error;
 
+	mutex_enter(&ztest_removal_lock);
```
Comment (ahrens) [inline cmd/ztest/ztest.c]: the old code was effectively a tryenter, do we want to use that here?

File: cmd/ztest/ztest.c [hunk 8]
```diff
@@ -533,10 +533,16 @@ _umem_logging_init(void)
 static void
 dump_debug_buffer(void)
 {
-	if (!ztest_dump_debug_buffer)
+	ssize_t ret __attribute__((unused));
```
Comment (ahrens) [inline cmd/ztest/ztest.c]: I think you can discard the return value with `(void)` as we do elsewhere

File: cmd/ztest/ztest.c [hunk 9]
```diff
@@ -866,7 +873,7 @@ process_options(int argc, char **argv)
 				usage(B_FALSE);
 			break;
 		case 'G':
-			ztest_dump_debug_buffer = B_TRUE;
+			zo->zo_dump_dbgmsg = 1;
```
Comment (ahrens) [inline cmd/ztest/ztest.c]: why did you change this from a boolean_t to an int?

File: module/zfs/zfs_debug.c [hunk 10]
```diff
@@ -212,11 +215,28 @@ __dprintf(const char *file, const char *func, int line, const char *fmt, ...)
 void
 zfs_dbgmsg_print(const char *tag)
 {
-	(void) printf("ZFS_DBGMSG(%s):\n", tag);
+	ssize_t ret __attribute__((unused));
```
Comment (ahrens) [inline module/zfs/zfs_debug.c]: `(void)`

File: lib/libzpool/kernel.c [hunk 11]
```diff
@@ -813,11 +819,37 @@ __dprintf(const char *file, const char *func, int line, const char *fmt, ...)
...
+
+		if (i < size) {
+			va_start(adx, fmt);
+			(void) vsnprintf(buf + i, size - i, fmt, adx);
+			va_end(adx);
+		}
+
+		/*
+		 * Get rid of trailing newline.
+		 */
```
Comment (ahrens) [inline lib/libzpool/kernel.c]: why is this needed?  are we zfs_dbgmsg()-ing things that have newlines?  Seems like we should fix that rather than changing the string after the fact (in some places but not others).

File: lib/libzpool/kernel.c [hunk 12]
```diff
@@ -813,11 +819,37 @@ __dprintf(const char *file, const char *func, int line, const char *fmt, ...)
...
+		nl = strrchr(buf, '\n');
+		if (nl != NULL)
+			*nl = '\0';
```
Comment (ahrens) [inline lib/libzpool/kernel.c]: This doesn't just get rid of trailing newlines, it gets rid of the last newline, which may have more text after it.

File: module/zfs/dsl_scan.c [hunk 13]
```diff
@@ -2386,6 +2386,20 @@ dsl_scan_ddt_entry(dsl_scan_t *scn, enum zio_checksum checksum,
...
+	 * that can add scan_io_t's to the vdev scan queues from
+	 * outside dsl_scan_sync(). For the most part this is ok
+	 * as long as it is called from within syncing context.
+	 * However, dsl_scan_sync() expects that no new sio's will
+	 * be added between when all the work for a scan is done
+	 * and the next txg when the scan is actually marked as
+	 * completed. This check ensures we do not sio's during
+	 * this period.
+	 */
+	if (scn->scn_done_txg != 0 && scn->scn_bytes_pending == 0)
```
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/zfs/dsl_scan.c]: This is to match the check in `dsl_scan_sync()`. We don't necessarily need that checkdepending on what we want. Having it there basically just ensures that if the scan is running we still scan incoming dedup blocks, even if the scan if finishing. Note that the "finishing" part of scanning can still take several  hours depending on the block sizes and RAM given to the dsl scan queues.

File: cmd/ztest/ztest.c (hunk 5 above)
Comment (tcaputi) [reply cmd/ztest/ztest.c]: This change was from @behlendorf. We were seeing several cases where reconstruction was taking a very long time. In fact, with 16 segments we were seeing ztest deadman timeouts. We wanted to reduce this workload so ztest has time to test other things so we lowered the max damaged segments to 12 and decreased how often they would occur.

File: module/zfs/vdev_indirect.c (hunk 6 above)
Comment (tcaputi) [reply module/zfs/vdev_indirect.c]: See comment above. with the existing code we can have 17 damaged segments which was causing the ztest deadman to trigger fairly often.

File: cmd/ztest/ztest.c (hunk 7 above)
Comment (tcaputi) [reply cmd/ztest/ztest.c]: This change was also from @behlendorf. I think the idea here was to make sure that this code path actually gets to run. As he mentioned, we might want to change this since a lot of these code paths are actually valid to race against each other in the actual kernel module.

File: cmd/ztest/ztest.c (hunk 8 above)
Comment (tcaputi) [reply cmd/ztest/ztest.c]: I tried that. GCC still complains. The only way to get it to shut up wass to save the return value. 🤷‍♂️

File: module/zfs/zfs_debug.c (hunk 10 above)
Comment (tcaputi) [reply module/zfs/zfs_debug.c]: See comment above.

File: lib/libzpool/kernel.c (hunk 11 above)
Comment (tcaputi) [reply lib/libzpool/kernel.c]: This was taken directly from the [zfs_debug.c code](https://github.com/zfsonlinux/zfs/blob/master/module/zfs/zfs_debug.c#L148-L208). I just wanted to ensure they behaved the same.

File: lib/libzpool/kernel.c (hunk 12 above)
Comment (tcaputi) [reply lib/libzpool/kernel.c]: See comment above.

File: cmd/ztest/ztest.c (hunk 9 above)
Comment (tcaputi) [reply cmd/ztest/ztest.c]: This was a minor thing. The variable in the struct is an int and I wanted to match `zo->zo_mmp_test`

File: module/zfs/dsl_scan.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/dsl_scan.c]: Which check in dsl_scan_sync()?  And what exactly do you mean by "finishing"?  Once scn_done_txg is set, could we still wait hours?

File: cmd/ztest/ztest.c (hunk 5 above)
Comment (ahrens) [reply cmd/ztest/ztest.c]: I see, so this isn't an issue of correctness (too much damage causing it to be unrecoverable), just performance (too much damage causing us to spend too much time doing successful recovery).

File: cmd/ztest/ztest.c (hunk 8 above)
Comment (ahrens) [reply cmd/ztest/ztest.c]: What's the error from gcc?  `(void)` works in other places in this file (at least, for other functions)

File: cmd/ztest/ztest.c (hunk 9 above)
Comment (ahrens) [reply cmd/ztest/ztest.c]: why did you change the variable from a boolean_t to an int?

File: lib/libzpool/kernel.c (hunk 11 above)
Comment (ahrens) [reply lib/libzpool/kernel.c]: Let's fix zfs_debug.c too then.

File: cmd/ztest/ztest.c (hunk 8 above)
Comment (behlendorf) [reply cmd/ztest/ztest.c]:
```
warning: ignoring return value of ‘ret’, declared with attribute warn_unused_result [-Wunused-result]
```

This a discussion which goes, way back https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425.  The short version is the behavior here isn't clearly defined by the standard and is compiler specific.

File: cmd/ztest/ztest.c (hunk 7 above)
Comment (behlendorf) [reply cmd/ztest/ztest.c]: We may, in fact as I mentioned in my comment,  think we should hold off entirely on the boolean -> lock conversion patch moment.  It's intention was to prevent racing between fault injection and removal, but we don't actually have clear evidence yet that it's needed.

File: cmd/ztest/ztest.c (hunk 5 above)
Comment (behlendorf) [reply cmd/ztest/ztest.c]: Yes, exactly.

File: module/zfs/dsl_scan.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/zfs/dsl_scan.c]: > Which check in dsl_scan_sync()?
```
else if (scn->scn_is_sorted && scn->scn_bytes_pending != 0) {
```

> And what exactly do you mean by "finishing"? Once scn_done_txg is set, could we still wait hours?

Yes. `scn_done_txg` is set when the last block of the scan is placed on one of the per-vdev scanning queues. The scan is not complete, however until all of the queued I/O has been issued. Back when I was testing this code, I saw this take about an hour on my test server, but I imagine it could be longer with other configurations. This check basically says "we won't add anything more to the queues after the normal scanning has completed adding its I/Os to the queues".

File: cmd/ztest/ztest.c (hunk 9 above)
Comment (tcaputi) [reply cmd/ztest/ztest.c]: The variable needed to be moved into `zo` so that it could properly be shared between processes. This struct did not have any boolean fields, even for fields such as `zo_mmp_test` which are clearly logically boolean. I assumed this might have something to do with ensuring alignment so I was just following the precedent.

File: lib/libzpool/kernel.c (hunk 11 above)
Comment (tcaputi) [reply lib/libzpool/kernel.c]: Ah. I think I see why this was done. `dprintf()` strings (by convention it seems) have a trailing newline while `zfs_dbgmsg()` strings do not. After looking around, I don't see any examples of multi-line calls to either of these functions. Not sure what the best way to handle this is....

File: cmd/ztest/ztest.c (hunk 8 above)
Comment (ahrens) [reply cmd/ztest/ztest.c]: Huh.  Any idea why all the other `(void)`'s in this file work fine, but it won't work here?

Comment (tcaputi) [reply cmd/ztest/ztest.c]: It has something to do with the `write()` call and its attributes specifically, I believe.

File: module/zfs/dsl_scan.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (ahrens) [reply module/zfs/dsl_scan.c]: I see.  I wasn't that familiar with the changes made due to sorted scan here.  Once `scn_done_txg` is set, we've found all the blocks to scan (presumably including dedup blocks).  Once that has happened, do we still need to do explicit scans in dsl_scan_ddt_entry()?  I don't think so.

File: cmd/ztest/ztest.c (hunk 8 above)
Comment (behlendorf) [reply cmd/ztest/ztest.c]: For several years now most of the standard glibc functions have had their prototypes updated to include the [warn_unused_result](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes) function attribute, which forces the warning.  Which is why we only see this in some places.

```c
#define __wur __attribute_warn_unused_result__

extern ssize_t write (int __fd, const void *__buf, size_t __n) __wur;
```

File: lib/libzpool/kernel.c (hunk 11 above)
Comment (tcaputi) [reply lib/libzpool/kernel.c]: Ok. I can do that.

File: module/zfs/dsl_scan.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/zfs/dsl_scan.c]: Ok. ill make the change.

**Comment by tcaputi** [comment]
@ahrens i believe your comments have been addressed.

File: cmd/ztest/ztest.c (hunk 5 above)
Comment (sdimitro) [inline cmd/ztest/ztest.c]: Do we want to maybe at to that comment the points discussed with Matt and Brian?
For example, something along those lines?
```
/*
 * Verify that even extensively damaged split blocks with many
 * ...
 * when reconstruction is known to be possible.
 *
 * Note: Keep in mind that the lower this value is, the more damage we inflict,
 * and the more time ztest spends in recovering that damage. We chose the
 * default of inducing damage 1/100th of the time so recovery is tested enough
 * but not too frequently to do so in the expense of other types of testing.
 */
```

File: lib/libzpool/kernel.c [hunk 14]
```diff
@@ -783,7 +783,8 @@ dprintf_setup(int *argc, char **argv)
  * =========================================================================
  */
 void
-__dprintf(const char *file, const char *func, int line, const char *fmt, ...)
+__dprintf(boolean_t dprint, const char *file, const char *func,
```
[the remark this answers is not in the document]
Comment (tcaputi) [reply lib/libzpool/kernel.c]: I did consider it, but in the kernel these functions are really almost exactly the same. Since the kernel implementation is the primary one, I wanted this code to match that.

File: module/zfs/txg.c [hunk 15]
```diff
@@ -562,22 +562,22 @@ txg_sync_thread(void *arg)
...
 		tx->tx_syncing_txg = txg;
 		DTRACE_PROBE2(txg__syncing, dsl_pool_t *, dp, uint64_t, txg);
-		txg_stat_t *ts = spa_txg_history_init_io(spa, txg, dp);
 		cv_broadcast(&tx->tx_quiesce_more_cv);
 
 		dprintf("txg=%llu quiesce_txg=%llu sync_txg=%llu\n",
 		    txg, tx->tx_quiesce_txg_waiting, tx->tx_sync_txg_waiting);
 		mutex_exit(&tx->tx_sync_lock);
 
+		txg_stat_t *ts = spa_txg_history_init_io(spa, txg, dp);
```
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/zfs/txg.c]: Good catch. I got the commit message backwards. I will fix it. I'm not sure what additional info you're looking for on the exact scenario. In general I think the idea of lock inversion is well understood enough to be recognizable as a bad thing when the code paths are pointed out. Is there something I'm missing?

File: module/zfs/zfs_debug.c [hunk 16]
```diff
@@ -144,15 +143,19 @@ __zfs_dbgmsg(char *buf)
 	mutex_exit(&zfs_dbgmsgs.pl_lock);
 }
 
+#ifdef _KERNEL
+
 void
-__dprintf(const char *file, const char *func, int line, const char *fmt, ...)
+__dprintf(boolean_t dprint, const char *file, const char *func,
```
[the remark this answers is not in the document]
Comment (tcaputi) [reply module/zfs/zfs_debug.c]: see above

**Comment by behlendorf** [comment]
Closing.  All but 53cb0bf were integrated.
